internal/mesh/internal/mappers/sidecarproxymapper/service_endpoints_mapper.go

```diff
@@ -0,0 +1,80 @@
+package sidecarproxymapper
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/cache/sidecarproxycache"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type Mapper struct {
+	cache *sidecarproxycache.Cache
+}
+
+func New(c *sidecarproxycache.Cache) *Mapper {
+	return &Mapper{
+		cache: c,
+	}
+}
+
+// MapServiceEndpointsToProxyStateTemplate maps catalog.ServiceEndpoints objects to the IDs of
+// ProxyStateTemplate.
+func (m *Mapper) MapServiceEndpointsToProxyStateTemplate(_ context.Context, _ controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	// This mapper needs to look up workload IDs from service endpoints and replace them with ProxyStateTemplate type.
+	var serviceEndpoints pbcatalog.ServiceEndpoints
+	err := res.Data.UnmarshalTo(&serviceEndpoints)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []controller.Request
+
+	// First, we need to generate requests from workloads this "endpoints" points to
+	// so that we can re-generate proxy state for the sidecar proxy.
+	for _, endpoint := range serviceEndpoints.Endpoints {
+		// Convert the reference to a workload to a ProxyStateTemplate ID.
+		// Because these resources are name and tenancy aligned, we only need to change the type.
+
+		// Skip service endpoints without target refs. These resources would typically be created for
+		// services external to Consul, and we don't need to reconcile those as they don't have
+		// associated workloads.
+		if endpoint.TargetRef != nil {
```

Comment:
I think you should do a type-conditional filter here to only consider targetRefs of type ServiceType. I know the validator prevents it, but it would help the mapper protect itself from weird input should we at some point in the future relax the validation rules a bit and forget about this code.

Reply:
Yeah I can add it. The target refs should be workload types in this case.

Reply:
After chatting with Matt K, I realized that I don't need to watch service endpoints at all, so this will go away in a future PR.